### PR TITLE
Adds progress bar for image push

### PR DESCRIPTION
**Pull request by shaheerkootteeri**
This PR resolves Issue #420 to have progress bar displayed for Image push.

### CODE REVIEW

File: pkg/imgpkg/bundle/contents.go [hunk 1]
```diff
@@ -30,7 +30,7 @@ type Contents struct {
 //go:generate go run github.com/maxbrunsfeld/counterfeiter/v6 . ImagesMetadataWriter
 type ImagesMetadataWriter interface {
 	ImagesMetadata
-	WriteImage(regname.Reference, regv1.Image) error
+	WriteImage(ref regname.Reference, img regv1.Image, _ chan regv1.Update) error
```
Comment (joaopapereira) [inline pkg/imgpkg/bundle/contents.go]:
```suggestion
	WriteImage(regname.Reference, regv1.Image, chan regv1.Update) error
```

File: pkg/imgpkg/plainimage/contents.go [hunk 2]
```diff
@@ -23,7 +23,7 @@ type Contents struct {
 }
 
 type ImagesWriter interface {
-	WriteImage(regname.Reference, regv1.Image) error
+	WriteImage(ref regname.Reference, reg regv1.Image, _ chan regv1.Update) error
```
Comment (joaopapereira) [inline pkg/imgpkg/plainimage/contents.go]:
```suggestion
	WriteImage(regname.Reference, regv1.Image, chan regv1.Update) error
```

File: pkg/imgpkg/registry/registry.go [hunk 3]
```diff
@@ -74,7 +74,7 @@ type ImagesReader interface {
 type ImagesReaderWriter interface {
 	ImagesReader
 	MultiWrite(imageOrIndexesToUpload map[regname.Reference]regremote.Taggable, concurrency int, updatesCh chan regv1.Update) error
-	WriteImage(regname.Reference, regv1.Image) error
+	WriteImage(ref regname.Reference, img regv1.Image, updatesCh chan regv1.Update) error
```
Comment (joaopapereira) [inline pkg/imgpkg/registry/registry.go]:
```suggestion
	WriteImage(regname.Reference, regv1.Image, chan regv1.Update) error
```

**Review by joaopapereira — COMMENTED** [review]
Made some comments on the PR it looks good. Let us try to see if we get these tests to pass

File: pkg/imgpkg/internal/util/progress_logger.go [hunk 4]
```diff
@@ -14,32 +14,40 @@ import (
...
 	ui                 goui.UI
 	finalMessage       string
 	errorMessagePrefix string
 }
 
-func (l *ProgressBarLogger) Start(progressChan <-chan regv1.Update) {
+// Start the display of the Progress Bar
+func (l *ProgressBarLogger) Start(ctx context.Context, progressChan <-chan regv1.Update) {
+	ctx, cancelFunc := context.WithCancel(ctx)
+	l.cancelFunc = cancelFunc
```
Comment (cppforlife) [inline pkg/imgpkg/internal/util/progress_logger.go:50]: im assuming we are not using ProgressBarLogger concurrently?

File: pkg/imgpkg/registry/with_progress.go [hunk 5]
```diff
@@ -4,6 +4,8 @@
 package registry
 
 import (
+	context "context"
```
Comment (cppforlife) [inline pkg/imgpkg/registry/with_progress.go]: nit: redundant aliasing

File: pkg/imgpkg/plainimage/contents.go [hunk 6]
```diff
@@ -18,12 +18,13 @@ import (
 )
 
 type Contents struct {
-	paths         []string
-	excludedPaths []string
+	paths           []string
+	excludedPaths   []string
+	displayProgress bool
```
Comment (cppforlife) [inline pkg/imgpkg/plainimage/contents.go]: dead code?

File: pkg/imgpkg/internal/util/progress_logger.go [hunk 7]
```diff
@@ -70,32 +78,43 @@ func (l *ProgressBarLogger) Start(progressChan <-chan regv1.Update) {
...
 }
 
+// End Write the final message
 func (l *ProgressBarNoTTYLogger) End() {
-	l.cancelFunc()
-	l.ui.BeginLinef(l.finalMessage)
+	if l.cancelFunc != nil {
+		l.cancelFunc()
+	}
+	if l.ui != nil && l.finalMessage != "" {
```
Comment (cppforlife) [inline pkg/imgpkg/internal/util/progress_logger.go:117]: do we need to check on `l.ui != nil` here?

File: pkg/imgpkg/internal/util/progress_logger.go (hunk 4 above)
Comment (shaheerkootteeri) [reply pkg/imgpkg/internal/util/progress_logger.go:50]: No we are not using

File: pkg/imgpkg/registry/with_progress.go (hunk 5 above)
Comment (shaheerkootteeri) [reply pkg/imgpkg/registry/with_progress.go]: I will remove the alias

File: pkg/imgpkg/plainimage/contents.go (hunk 6 above)
Comment (shaheerkootteeri) [reply pkg/imgpkg/plainimage/contents.go]: we wouldn't need the displayProgress flag as it is not used anymore, however we would need the paths and excluded paths. I will remove the displayProgress flag

File: pkg/imgpkg/internal/util/progress_logger.go (hunk 7 above)
Comment (shaheerkootteeri) [reply pkg/imgpkg/internal/util/progress_logger.go:117]: In the case of a `NOOP progress logger `as part of the bundle push , we are not passing any UI as there will not be any progress bar displaying. However we could have the same accepting ui as well and we would need the `nil` check to decide if possible to display final message.

File: pkg/imgpkg/plainimage/contents.go (hunk 6 above)
Comment (shaheerkootteeri) [reply pkg/imgpkg/plainimage/contents.go]: Removed the displayProgress as part of the latest commit

File: pkg/imgpkg/registry/with_progress.go (hunk 5 above)
Comment (shaheerkootteeri) [reply pkg/imgpkg/registry/with_progress.go]: Addressed in the latest commit.

**Review by joaopapereira — APPROVED** [review]
LGTM

**Pull request merged**